## ANOVA

In [8]:

import numpy as np
import pandas as pd
import random 

import matplotlib.pyplot as plt
import seaborn as sns

import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler

import warnings
warnings.filterwarnings("ignore")

plt.rcParams["figure.figsize"] = (10,8) 

En el pair programming de hoy usaremos el set de datos que guardamos en el pair programming de normalización y estandarización.

En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

- Hacemos el ANOVA como práctica, pero nuestra variable respuesta no cumple la asunción de normalidad, por lo que no podemos hacer regresión lineal

In [9]:
#Cargamos el dataframe con el que vamos a trabajar.
df = pd.read_csv("../ficheros/diamons_estandarizados.csv", index_col = 0)
df.head(2)

carat      cut color clarity     depth     table  price         x  \
0 -0.746032    Ideal     E     SI2 -0.214286 -0.666667    326 -0.931818   
1 -0.777778  Premium     E     SI1 -1.428571  1.333333    326 -0.965909   

         y         z  
0 -0.91954 -0.909091  
1 -1.00000 -1.018182

In [10]:
# Iniciamos el ANOVA construyendo la fórmula con la variable respuesta y las predictoras
lm = ols("price ~ carat + cut + color + clarity + depth + table + x + y + z", data= df).fit()
sm.stats.anova_lm(lm)

df        sum_sq       mean_sq              F         PR(>F)
cut           4.0  4.734489e+09  1.183622e+09    1967.220812   0.000000e+00
color         6.0  8.471829e+09  1.411971e+09    2346.745167   0.000000e+00
clarity       7.0  9.245534e+09  1.320791e+09    2195.199369   0.000000e+00
carat         1.0  3.302812e+11  3.302812e+11  548938.653540   0.000000e+00
depth         1.0  9.179761e+05  9.179761e+05       1.525708   2.167635e-01
table         1.0  3.877882e+07  3.877882e+07      64.451739   1.010557e-15
x             1.0  6.445927e+08  6.445927e+08    1071.335282  1.586443e-232
y             1.0  4.895838e+07  4.895838e+07      81.370516   1.935488e-19
z             1.0  1.052687e+08  1.052687e+08     174.960124   7.123557e-40
Residual  50218.0  3.021478e+10  6.016723e+05            NaN            NaN

Según vemos aquí, todas nuestras variables tienen un p-valor menor de 0.05, por lo que debemos rechazar la hipótesis nula, lo que quiere decir que todas las variables influyen en nuestra variable respuesta, que es el precio de los diamantes.

In [11]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.921
Model:                            OLS   Adj. R-squared:                  0.921
Method:                 Least Squares   F-statistic:                 2.555e+04
Date:                Wed, 25 Jan 2023   Prob (F-statistic):               0.00
Time:                        19:12:38   Log-Likelihood:            -4.0558e+05
No. Observations:               50242   AIC:                         8.112e+05
Df Residuals:                   50218   BIC:                         8.114e+05
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         -175.6187     36.501     -4.811      0.000    -247.160    -104.077
cut[T.Good]        406.0150     24.325     16.691      0.000     358.338     453.692
cut[T.Ideal]       620.6655     24.118     25.735      0.000     573.395     667.936
cut[T.Premium]     587.7553     23.043     25.507      0.000     542.590     632.920
cut[T.Very Good]   521.6468     23.561     22.140      0.000     475.467     567.826
color[T.E]        -162.2045     12.547    -12.928      0.000    -186.797    -137.612
color[T.F]        -211.3518     12.742    -16.587      0.000    -236.327    -186.377
color[T.G]        -349.9169     12.506    -27.980      0.000    -374.429    -325.405
color[T.H]        -708.3351     13.364    -53.004      0.000    -734.528    -682.142
color[T.I]       -1123.8565     15.173    -74.070      0.000   -1153.596   -1094.117
color[T.J]       -1767.0199     18.953    -93.230      0.000   -1804.168   -1729.871
clarity[T.IF]     3960.5498     36.252    109.250      0.000    3889.495    4031.604
clarity[T.SI1]    2757.9999     30.935     89.155      0.000    2697.367    2818.632
clarity[T.SI2]    2001.1528     31.063     64.423      0.000    1940.269    2062.036
clarity[T.VS1]    3492.8409     31.601    110.531      0.000    3430.903    3554.779
clarity[T.VS2]    3226.2761     31.112    103.700      0.000    3165.297    3287.255
clarity[T.VVS1]   3815.7967     33.378    114.322      0.000    3750.376    3881.217
clarity[T.VVS2]   3828.3753     32.487    117.843      0.000    3764.700    3892.050
carat             5992.3580     34.741    172.487      0.000    5924.265    6060.451
depth               69.4126     10.456      6.638      0.000      48.918      89.907
table              -55.2189      6.259     -8.822      0.000     -67.487     -42.951
x                -1493.0495    134.869    -11.070      0.000   -1757.394   -1228.705
y                 1972.0205    136.714     14.424      0.000    1704.059    2239.982
z                -1737.2436    131.338    -13.227      0.000   -1994.668   -1479.819
==============================================================================
Omnibus:                    13645.432   Durbin-Watson:                   1.094
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           104730.262
Skew:                           1.098   Prob(JB):                         0.00
Kurtosis:                       9.724   Cond. No.                         63.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- Nuestro *R cuadrado* es 0.921, el 92 % de lo que cambia la variable respuesta se debe a nuestras variables predictoras.

- Respecto al *error estándar*, los que tienen el valor más bajo son **Depth** y **Table**, lo que quiere decir que la estimación de la variable respuesta será más precisa. El **color** también parece bastante preciso. Sin embargo, tanto x como y tienen valores muy altos.

- Respecto a la *P*, son todas  menores de 0.05, con lo cual son importantes al hacer las predicciones. 
    
- Respecto al *coeficiente*, vemos que sobre todo el **carat (quilates)** tiene un coeficiente muy alto. Cuanto mayor son los quilates, el precio de los diamantes sube mucho. En el caso de **depth** es positivo y con un valor bajo y **table** la relación es inversamente proporcional.

Hacemos lo mismo con los datos sin estandarizar, para comprobar los resultados

In [12]:
df2= pd.read_csv("../ficheros/diamons_01.csv", index_col = 0)
df2.head(2)

carat      cut color clarity  depth  table  price     x     y     z
1   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
2   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [13]:
lm2 = ols("price ~ carat + cut + color + clarity + depth + table + x + y + z", data= df2).fit()
sm.stats.anova_lm(lm2)

df        sum_sq       mean_sq              F         PR(>F)
cut           4.0  4.734489e+09  1.183622e+09    1967.220812   0.000000e+00
color         6.0  8.471829e+09  1.411971e+09    2346.745167   0.000000e+00
clarity       7.0  9.245534e+09  1.320791e+09    2195.199369   0.000000e+00
carat         1.0  3.302812e+11  3.302812e+11  548938.653540   0.000000e+00
depth         1.0  9.179761e+05  9.179761e+05       1.525708   2.167635e-01
table         1.0  3.877882e+07  3.877882e+07      64.451739   1.010557e-15
x             1.0  6.445927e+08  6.445927e+08    1071.335282  1.586443e-232
y             1.0  4.895838e+07  4.895838e+07      81.370516   1.935488e-19
z             1.0  1.052687e+08  1.052687e+08     174.960124   7.123557e-40
Residual  50218.0  3.021478e+10  6.016723e+05            NaN            NaN

In [15]:
lm2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.921
Model:                            OLS   Adj. R-squared:                  0.921
Method:                 Least Squares   F-statistic:                 2.555e+04
Date:                Wed, 25 Jan 2023   Prob (F-statistic):               0.00
Time:                        19:17:31   Log-Likelihood:            -4.0558e+05
No. Observations:               50242   AIC:                         8.112e+05
Df Residuals:                   50218   BIC:                         8.114e+05
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept        -5013.5990    510.137     -9.828      0.000   -6013.473   -4013.725
cut[T.Good]        406.0150     24.325     16.691      0.000     358.338     453.692
cut[T.Ideal]       620.6655     24.118     25.735      0.000     573.395     667.936
cut[T.Premium]     587.7553     23.043     25.507      0.000     542.590     632.920
cut[T.Very Good]   521.6468     23.561     22.140      0.000     475.467     567.826
color[T.E]        -162.2045     12.547    -12.928      0.000    -186.797    -137.612
color[T.F]        -211.3518     12.742    -16.587      0.000    -236.327    -186.377
color[T.G]        -349.9169     12.506    -27.980      0.000    -374.429    -325.405
color[T.H]        -708.3351     13.364    -53.004      0.000    -734.528    -682.142
color[T.I]       -1123.8565     15.173    -74.070      0.000   -1153.596   -1094.117
color[T.J]       -1767.0199     18.953    -93.230      0.000   -1804.168   -1729.871
clarity[T.IF]     3960.5498     36.252    109.250      0.000    3889.495    4031.604
clarity[T.SI1]    2757.9999     30.935     89.155      0.000    2697.367    2818.632
clarity[T.SI2]    2001.1528     31.063     64.423      0.000    1940.269    2062.036
clarity[T.VS1]    3492.8409     31.601    110.531      0.000    3430.903    3554.779
clarity[T.VS2]    3226.2761     31.112    103.700      0.000    3165.297    3287.255
clarity[T.VVS1]   3815.7967     33.378    114.322      0.000    3750.376    3881.217
clarity[T.VVS2]   3828.3753     32.487    117.843      0.000    3764.700    3892.050
carat             9511.6794     55.144    172.487      0.000    9403.596    9619.763
depth               49.5804      7.469      6.638      0.000      34.941      64.220
table              -18.4063      2.086     -8.822      0.000     -22.496     -14.317
x                 -848.3236     76.630    -11.070      0.000    -998.519    -698.128
y                 1133.3451     78.571     14.424      0.000     979.344    1287.346
z                -1579.3124    119.398    -13.227      0.000   -1813.335   -1345.290
==============================================================================
Omnibus:                    13645.432   Durbin-Watson:                   1.094
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           104730.262
Skew:                           1.098   Prob(JB):                         0.00
Kurtosis:                       9.724   Cond. No.                     1.28e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.28e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Los datos nos salen igual que en los estandarizados, por lo que no lo guardamos